packages/data/src/options/selectors.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * External Dependencies
+ */
+
+import { select } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { STORE_NAME } from './constants';
+
+/**
+ * Get options from state tree.
+ *
+ * @param {Object} state - Reducer state
+ * @param {Array} names - Option names
+ */
+export const getOptions = ( state, names ) => {
+	// if the options aren't in state, then call on the selector that has a resolver.
+	if ( names.some( ( name ) => ! state[ name ] ) ) {
+		return select( STORE_NAME ).getOptionsWithRequest( names );
+	}
+
+	return names.reduce( ( result, name ) => {
+		result[ name ] = state[ name ];
+		return result;
+	}, {} );
+};
+
+/**
+ * Get options from state tree or make a request if unresolved.
+ *
+ * @param {Object} state - Reducer state
+ * @param {Array} names - Option names
+ */
+export const getOptionsWithRequest = ( state, names ) => {
+	return names.reduce( ( result, name ) => {
+		result[ name ] = state[ name ];
+		return result;
+	}, {} );
+};
+
+/**
+ * Determine if options are being requested.
+ *
+ * @param {Object} state - Reducer state
+ */
+export const isGetOptionsRequesting = ( state ) => {
```

Comment:
I think we can replace these additional selectors with those built in to wp data.  Outlined an issue with more detail [here](https://github.com/woocommerce/woocommerce-admin/issues/4384).

Reply:
Thats an excellent idea! The way this is setup is `getOptions` calls resolver `getOptionsWithRequest` if the option hasn't been preloaded, which leaves us with this:

```js
const options = getOptions( args );
const isRequesting = ( 'getOptionsWithRequest', [ args ] ) );
// What we really want is this:
// const isRequesting = ( 'getOptions', [ args ] ) );
```
More on why its that way in p90Yrv-1uN#comment-3842.

One option is to introduce `isOptionsResolving` to map resolver names. This is less than ideal because I'd rather hide implementation details from the consumer.

```js
export const isOptionsResolving = ( state, selectorName, args ) => {
	return select( STORE_NAME ).isResolving(
		selectorName === 'getOptions' ? 'getOptionsWithRequest' : selectorName,
		args
	);
};
```
Any other ideas?

Reply:
>Another option would be to get rid of getOptions all together and just have any request for data always go through getOption to simplify the code.

I think I like this idea the best.  It's a simple and easy to understand API and fixes issues with caching or needing to call with the exact same arguments used to determine resolution.

I originally wrote `getOptions` with an array for the purpose of avoiding multiple trips to the server when multiple options were needed, but if we wanted to keep this from a performance perspective, maybe we could collect requests for options and then fetch at the end of the stack.

For example, assume option A is cached but B and C need to be fetched and `optionsToRequest` and `fetches` are available across calls.

```js
// in component
const a = getOption( 'a' );
const b = getOption( 'b' );
const c = getOption( 'c' );

// in resolvers, store a persisted array of items that need to be requested
optionsToRequest.push( optionName );

// Use a timeout of 1ms to push to the end of the stack
setTimeout( () => {
  // optionsToRequest should now be ['b', 'c']
  // Avoid fetching multiple times for a group of options.
  fetchPromise = fetches[ optionsToRequest.join(',') ]
  if ( fetchStarted ) {
     yield fetchPromise;
     return;
  }

  // fetch options in array optionsToRequest
  fetches[ optionsToRequest.join(',') ]  = yield apiFetch( {
    path: url,
    method: 'GET',
    data: optionsToRequest
   } );
  
}, 1 );
```

Do you think something like that is feasible?

Reply:
Nice one @joshuatf !

I like the request collection technique. The thing that deterred me from this solution was it complicating the hydration part because I'll need to loop through all of the options individually. I'll look for solution to that one though.